packages/core/src/transactions/utils/transactions.ts

```diff
@@ -1,212 +1,314 @@
-//     .methods['forward']
-
-//   return (forwarderAddress, script) => ({
-//     ...directTransaction, // Options are overwriten by the values below
-//     from: sender,
-//     to: forwarderAddress,
-//     data: forwardMethod(script).encodeABI(),
-//   })
-// }
-
-// export async function getRecommendedGasLimit(
-//   web3,
-//   estimatedGasLimit,
-//   { gasFuzzFactor = DEFAULT_GAS_FUZZ_FACTOR } = {}
-// ) {
-//   const latestBlock = await web3.eth.getBlock('latest')
-//   const latestBlockGasLimit = latestBlock.gasLimit
-
-//   const upperGasLimit = Math.round(
-//     latestBlockGasLimit * PREVIOUS_BLOCK_GAS_LIMIT_FACTOR
-//   )
-//   const bufferedGasLimit = Math.round(estimatedGasLimit * gasFuzzFactor)
-
-//   if (estimatedGasLimit > upperGasLimit) {
-//     // TODO: Consider whether we should throw an error rather than returning with a high gas limit
-//     return estimatedGasLimit
-//   } else if (bufferedGasLimit < upperGasLimit) {
-//     return bufferedGasLimit
-//   } else {
-//     return upperGasLimit
-//   }
-// }
+import ethers from 'ethers'
+
+import TransactionRequest from '../TransactionRequest'
+import App from '../../wrappers/App'
+import { Abi, FunctionFragment } from '../../types'
+
+const DEFAULT_GAS_FUZZ_FACTOR = 1.5
+const PREVIOUS_BLOCK_GAS_LIMIT_FACTOR = 0.95
+
+const erc20ABI = [
+  'function balanceOf(address _who) public view returns (uint256)',
+  'function allowance(address _owner, address _spender) public view returns (uint256)',
+  'function approve(address _spender, uint256 _value) public returns (bool)',
+]
+
+const forwarderAbi = [
+  'function forward(bytes evmCallScript) public',
+  'function canForward(address sender, bytes evmCallScript) public view returns (bool)',
+  'function isForwarder() external pure returns (bool)',
+]
+
+const forwarderFeeAbi = [
+  'function forwardFee() external view returns (address, uint256)',
+]
+
+// Is the given method a full signature, e.g. 'foo(arg1,arg2,...)'
+export const isFullMethodSignature = (methodSignature: string): boolean => {
+  return Boolean(methodSignature) &&
+  methodSignature.includes('(') &&
+  methodSignature.includes(')')
+}
+
+export interface directTransaction extends TransactionRequest {
+  token?: {
+    address: string,
+    value: string
+    spender: string
+  }
+  pretransaction?: {
+    from: string,
+    to: string
+    data: string
+  }
+}
+
+export async function applyPretransaction(directTransaction: directTransaction) {
+  // Token allowance pretransaction
+  if (directTransaction.token) {
+    const {
+      from,
+      to,
+      token: { address: tokenAddress, value: tokenValue, spender },
+    } = directTransaction
+
+    // Approve the transaction destination unless an spender is passed to approve a different contract
+    const approveSpender = spender || to
+
+    // TODO: handle provider network: using chainId from Transaction Request?
+    const provider = ethers.getDefaultProvider()
```

Comment:
This is one of my biggest unknowns. How we would like to get the provider. I think we can get the `chainId` and some point and use that to call `getDefaultProvider`

Reply:
I was wondering if it would make sense to pass the provider to `Intent#transactions(provider)` and `Intent#paths(provider)`? Or should we have a global provider set during the initialization, separated from the account address?

Reply:
I wonder if this is a requirement or we are able to infer it internally. But maybe it is an antipattern to do that. Regardless I believe asking at the method level is better than having as a global requirement.

Reply:
> I wonder if this is a requirement or we are able to infer it internally. But maybe it is an antipattern to do that.

Yeah we could try to infer it, but it would remove the possibility for app users to pick the provider they want (e.g. through [useWallet()](https://github.com/aragon/use-wallet/)) and not necessarily the injected one.

> Regardless I believe asking at the method level is better than having as a global requirement.

Cool, we might also want to add the possibility to define a global provider in `connect()`, that could be overridden at the method level?